terraform/node_output.go

```diff
@@ -9,6 +9,98 @@ import (
 	"github.com/hashicorp/terraform/lang"
 )
 
+// NodePlannableOutput is the placeholder for an output that has not yet had
+// its module path expanded.
+type NodePlannableOutput struct {
+	Addr   addrs.OutputValue
+	Module addrs.Module
+	Config *configs.Output
+}
+
+var (
+	_ GraphNodeSubPath       = (*NodePlannableOutput)(nil)
+	_ RemovableIfNotTargeted = (*NodePlannableOutput)(nil)
+	_ GraphNodeReferenceable = (*NodePlannableOutput)(nil)
+	//_ GraphNodeEvalable          = (*NodePlannableOutput)(nil)
+	_ GraphNodeReferencer        = (*NodePlannableOutput)(nil)
+	_ GraphNodeDynamicExpandable = (*NodePlannableOutput)(nil)
+)
+
+func (n *NodePlannableOutput) DynamicExpand(ctx EvalContext) (*Graph, error) {
+	var g Graph
+	expander := ctx.InstanceExpander()
+	for _, module := range expander.ExpandModule(ctx.Path().Module()) {
+		o := &NodeApplyableOutput{
+			Addr:   n.Addr.Absolute(module),
+			Config: n.Config,
+		}
+		// log.Printf("[TRACE] Expanding output: adding %s as %T", o.Addr.String(), o)
+		g.Add(o)
+	}
+	return &g, nil
+}
+
+func (n *NodePlannableOutput) Name() string {
+	return n.Addr.AbsString(n.Module)
+}
+
+// GraphNodeSubPath
+func (n *NodePlannableOutput) Path() addrs.ModuleInstance {
+	// Return an UnkeyedInstanceShim as our placeholder,
+	// given that modules will be unexpanded at this point in the walk
+	return n.Module.UnkeyedInstanceShim()
+}
+
+// GraphNodeReferenceable
+func (n *NodePlannableOutput) ReferenceableAddrs() []addrs.Referenceable {
+	// An output in the root module can't be referenced at all.
+	if n.Module.IsRoot() {
+		return nil
+	}
+
+	// the output is referenced through the module call, and via the
+	// module itself.
+	_, call := n.Module.Call()
+
+	// FIXME: make something like ModuleCallOutput for this type of reference
+	// that doesn't need an instance shim
```

Comment:
This is likely very important to get planned outputs working. Maybe this type should be `ModuleInstanceCallOutput` and `ModuleCallOutput` is the unexpanded version?